### PR TITLE
network: support reconfiguring netdev

### LINKED ISSUE #9627: changing .netdev files and restarting has no effect if interface already exists
**systemd version the issue has been seen with**
> 239-5

**Used distribution**
 > debian stretch & debian sid

**Expected behaviour you didn't see**
 > when restarting systemd-networkd, the vlan does not reconfigured. When changing the vlan id in the .netdev file and restarting, the vlan on the interface stays the same.

**Steps to reproduce the problem**

setup a vlan using systemd-networkd
restart systemd-networkd and confirm vlan is created (ip -d link show)
change vlan id in the .netdev file of the vlan
restart systemd-networkd
observe the vlan id on the interface is still the intial vlan id and not the new one (ip -d link show)

### LINKED ISSUE #27177: network: make 'networkctl reload' more graceful
### Component

systemd-networkd

### Is your feature request related to a problem? Please describe

- also reconfigure netdevs if .netdev files are updated (c.f. #9627 and #22557),
- keep DHCP address if possible,
- add dry-run mode (reload, but do not apply unless `networkctl reconfigure` is called),

and so on.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

_No response_

### LINKED ISSUE #34907: systemd-networkd fails to reload .netdev units
### systemd version the issue has been seen with

255

### Used distribution

Fedora 40

### Linux kernel version used

6.11.4-201.fc40.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

I got a VLAN interface defined by a `.netdev` unit. I want to effectively "change" the VLAN ID by deleting and recreating the interface.

For example, a VLAN interface named `myvlan` exists. Then:
1. Modify `Id=2` to `Id=3` under the `[VLAN]` section of its `myvlan.netdev` file.
2. `networkctl delete myvlan`
3. `networkctl reload`

I would expect the VLAN interface to have been deleted, recreated, and now under VLAN ID 3.

In systemd v255 nor v256, neither happens. The VLAN interface is not recreated.
In systemd v249, both happen. The interface is recreated under VLAN ID 3.

I've not tested intermediate releases.

### Unexpected behaviour you saw

Two issues:
1. Issuing `networkctl reload` does not recreate the `myvlan` interface. By executing `touch` on the parent's `.network` file, a subsequent `networkctl reload` will however recreate the interface.
3. The recreated VLAN interface is on VLAN ID 2, when it should've been on 3.

### Steps to reproduce the problem

Basic configuration of an Ethernet device and a VLAN interface:
```ini
root@cat:/run/systemd/network# cat enp0s31f6.network 
[Match]
Name=enp0s31f6

[Network]
VLAN=myvlan
```
```ini
root@cat:/run/systemd/network# cat myvlan.netdev 
[NetDev]
Name=myvlan
Kind=vlan

[VLAN]
Id=2
```
```ini
root@cat:/run/systemd/network# ip -d l sh myvlan
5: myvlan@enp0s31f6: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 30:5a:3a:45:e5:b5 brd ff:ff:ff:ff:ff:ff promiscuity 0 allmulti 0 minmtu 0 maxmtu 65535 
    vlan protocol 802.1Q id 2 <REORDER_HDR> numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 gso_ipv4_max_size 65536 gro_ipv4_max_size 65536 
```

The interface `myvlan` exists exactly as expected.

Issue 1:

```ini
root@cat:/run/systemd/network# sed -i 's/Id=2/Id=3/' myvlan.netdev
root@cat:/run/systemd/network# networkctl delete myvlan
root@cat:/run/systemd/network# networkctl reload
root@cat:/run/systemd/network# ip -d l sh myvlan
Device "myvlan" does not exist.
```

Issue 2:
```ini
root@cat:/run/systemd/network# touch enp0s31f6.network
root@cat:/run/systemd/network# networkctl reload
root@cat:/run/systemd/network# ip -d l sh myvlan
6: myvlan@enp0s31f6: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 30:5a:3a:45:e5:b5 brd ff:ff:ff:ff:ff:ff promiscuity 0 allmulti 0 minmtu 0 maxmtu 65535 
    vlan protocol 802.1Q id 2 <REORDER_HDR> numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 gso_ipv4_max_size 65536 gro_ipv4_max_size 65536 
```

Observe how, even though `myvlan.netdev` specifies `Id=3` and the interface didn't exist before, it is created with VLAN ID 2.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Closes #9627
Closes #27177.
Closes #34907.
Replaces #22557.

### CODE REVIEW

**Comment by yuwata** [comment]
TODO: do not trigger assertions for netdev->manager in callbacks.

File: man/networkctl.xml [hunk 1]
```diff
@@ -413,14 +413,22 @@ s - Service VLAN, m - Two-port MAC Relay (TPMR)
...
-        Note that even if an existing <filename>.netdev</filename> is modified or removed,
-        <command>systemd-networkd</command> does not update or remove the netdev.
-        If a new, modified or removed <filename>.network</filename> file is found, then all interfaces
-        which match the file are reconfigured.</para>
+        <listitem>
+          <para>Reload <filename>.netdev</filename> and <filename>.network</filename> files.</para>
 
-        <xi:include href="version-info.xml" xpointer="v244"/></listitem>
+          <para>If a new or modified <filename>.netdev</filename> file is found, then the corresponding
+          netdev is created or updated, respectively. Note, if the corresponding interface is already exist,
```
Comment (bluca) [inline man/networkctl.xml]: "already exists"

File: man/networkctl.xml [hunk 2]
```diff
@@ -413,14 +413,22 @@ s - Service VLAN, m - Two-port MAC Relay (TPMR)
...
+          then some of new settings may not be applied. E.g., VLAN ID cannot be changed after created, so
```
Comment (bluca) [inline man/networkctl.xml]: "after the interface was created"

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@bluca Thank you for the review. Addressed the comments. Upgrading the green label.

**Pull request merged**